### ISSUE TITLE
add "show value on hover" to controls using tooltip

### ISSUE THREAD

**Issue by jcbeppler** [swharden/ScottPlot#410]
First of all I have to say congratulations for the great job done! Your library is not only growing fast but it is also incredibility fast, flexible and unique. :)

I have some ideas and I know they are not very simple to do (I attached a drawing, I believe it's easier to explain) but let's go for them:

1- It would be nice if the hover feature could be built-in and just an enable/disable feature. If enabled, a box with the nearest values could show up for all the lines. This would save a lot of copy paste code if people have multiple charts.

2- The legend could be moved to the left side, the "Vertical Units" could go down and, together with the legend, the mouse points could be displayed together. On the left side it won't increase the needed size too much because there's already size reserved for the vertical units. The line color representation on the legend could also be reduced to not occupy too much space. 

3- The horizontal units could also be moved to the left and up, so it doesn't use too much space on the bottom, this is particularly useful if you have multiple charts one under the other  (of course, this is really not needed, maybe could have a way of choosing).

For the mouse hover, you will have multiple lines which not necessarily have the same X-axis, so it could have something like a "delta", which could be specified. It will show just the points if the line point is inside this range.

![ScottPlot_Hover_Suggestion](https://user-images.githubusercontent.com/59994318/82308099-996cbe80-99c1-11ea-9238-2bc7a7ea97c9.png)

**Comment by swharden** [comment]
Hey @jcbeppler thanks for your suggestions! Here are my thoughts. I'll rename this issue to focus on your first suggestion and create a new issue to track the third.

### Show value under cursor

> It would be nice if the hover feature could be built-in and just an enable/disable feature

I agree! This is a bit cumbersome right now. Instead of drawing on the graph (bitmap) what do you think about showing a mouse tooltip? This might make the task a lot easier to implement. NPlot does something like this (pictured)

![image](https://user-images.githubusercontent.com/4165489/82332458-6a4f5080-99b3-11ea-8a67-6aec85af8d4b.png)

### Legend outside plot area

This is a good idea which has been recently discussed #388 and is currently on the [ScottPlot roadmap](https://github.com/swharden/ScottPlot/blob/master/dev/roadmap.md) as a feature I home to implement soon.

### User specified alignment of axis labels

This idea seems useful in general. Title too! I'll consider the best way to implement this.

**Comment by jcbeppler** [comment]
Hello,

A tooltip is a great idea, sounds good! I think the most important is to be able to see more than one point value at a time in some range in case you have multiple curves. E.g. sometimes you have 5 curves and 2 has the same value, so you see 4 curves and never know under which one is the missing curve...


I tested the LiveCharts library a couple of months ago, interesting but quite slow for high amount of points... I think the developers gave up on the project around 2 years ago, but maybe it's nice to get some ideas for this part of showing values.
https://lvcharts.net/App/examples/v1/wpf/Series

**Comment by swharden** [comment]
Hey @jcbeppler, if there are two lines with overlapping data, and the mouse hovers over them, what should get displayed? I'm leaning toward always displaying data values from the "uppermost" plot (whichever plot was plotted last, ending up on the top of the image).

It's funny you mentioned LiveCharts - I'm working on a review of all the currently-available charting libraries and I just finished looking over it ([link / incomplete article](https://swharden.com/CsharpDataVis/plotting/livecharts.md.html)). The library is indeed slow, and the project seems to have died around 2018. In reviewing lots of plotting libraries ([full list / incomplete page](https://swharden.com/CsharpDataVis/#plotting-data-with-c)) it's interesting to see what they have in common and where they differ...

**Comment by swharden** [comment]
... displaying one number when hovering over two curves that overlap at the same point also makes sense because, if they're overlapping, the value should probably be about the same anyway 🤔

**Comment by jcbeppler** [comment]
Hello, I meant is nice to have the curve names together with the displayed value and, regarding one curve one over the other, I made the drawing bellow to better explain. In that case, there are 3 lines: A, B, C... B and C have the same value but if you don't have the tooltip with the names (or the values on the legend), you never know where Line B is... it could be under A, C, Sin, or Cos... 

![ScottPlot_Hover_Suggestion2](https://user-images.githubusercontent.com/59994318/82485715-25730900-9adc-11ea-9de0-0f1bc0d8a622.png)

Also when analyzing data, is much easier if you can compare the value of multiple lines in the same (or almost the same) instant of time (X-Axis) in case you are doing data acquisition.

I really liked your library review. I'm not very experienced with WForms, I worked more with WPF/UWP with the LiveCharts library. Bellow is one example with it on what I think would be great to achieve (maybe it's possible to do some copy/paste...):

![ScottPlot_Hover_LiveCharts](https://user-images.githubusercontent.com/59994318/82487641-f5793500-9ade-11ea-9c59-fbef4c2cbc00.png)

Of course, this is just a suggestion, I have no idea how much effort is needed to do that.

**Comment by swharden** [comment]
Thank you for clarifying @jcbeppler, I understand now. 

I think the best way to create a new interface for plottables capable of showing the value under the cursor. Something like `IHoverValue`. It would have a method like like

```cs
(double xVal, double yVal) ValueAtPixel(double xPx, double yPx, double snapX = 3, double snapY = 3);
```

where `xPx` and `yPx` is the mouse position, and `snapX` and `snapY` are how close the mouse has to be to be considered a "hover".

To get the value for every plottable beneath a vertical line (what your screenshot shows) we would just set `snapY` to `double.PositiveInfinity`.

I'll think about this some more!

**Comment by swharden** [comment]
After #414 (#415) is figured-out we can create a new "show value on hover" demo that uses it, addressing this issue's goal.

**Comment by jcbeppler** [comment]
Hi @swharden , that sounds great!

**Comment by swharden** [comment]
hey @jcbeppler, the new `PlotScatterHighlight()` (#414) should help here, and I also added a "show value under cursor" tooltip option to the windows forms control (though it's a bit flickery). I updated the "show value under mouse" WinForms demo too, so you have some more options for how to display the value under the cursor now. Thanks for the suggestion!
